jscomp/runtime/caml_js_exceptions.res

```diff
@@ -23,17 +23,32 @@
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */
 
 exception Error = JsError
-type js_error = {cause: exn}
-/**   
-   This function has to be in this module Since 
-   [Error] is defined here 
-*/
-let internalToOCamlException = (e: unknown) =>
-  if Caml_exceptions.is_extension((Obj.magic(e): js_error).cause) {
-    (Obj.magic(e): js_error).cause
+
+let internalAnyToExn = (any: 'a): exn =>
+  if Obj.magic(any) && Js.typeof(Obj.magic(any)["RE_EXN_ID"]) === "string" {
+    any->Obj.magic
   } else {
-    JsError(e)
+    {
+      "RE_EXN_ID": "JsError",
+      "_1": any,
+    }->Obj.magic
+  }
+
+%%raw(`class RescriptError extends Error {
+  constructor(message) {
+    super(message);
+    this.RE_EXN_ID = message;
   }
+}`)
+
+@new
+external internalMakeExn: string => exn = "RescriptError"
+// Reassign it here from external to let, since RescriptError is not exported
+let internalMakeExn = internalMakeExn
+
+let internalFromExtension = (_ext: 'a): exn => {
+  %raw(`Object.assign(new RescriptError(_ext.RE_EXN_ID), _ext)`)
+}
```

Comment:
I think it can be optimised to store the whole payload in an error field instead of spreading the object. But it requires modifying the matching logic, so I've done the easiest thing here to have at least something working

Reply:
Couldn't we just `new RescriptError(id, payload)` ?

Reply:
It'll still require to do `Object.assign` inside of the `RescriptError` constructor. Until I updated matching logic, it's not possible to get rid of `Object.assign`